### PR TITLE
Fixes for 32-bit SPARC on Linux

### PR DESCRIPTION
**Pull request by glaubitz**
This PR fixes a number of issues which previously prevented `rustc` from being built
successfully for 32-bit SPARC using the `sparc-unknown-linux-gnu` triplet.

In particular, it adds linking against `libatomic` where necessary, uses portable `AtomicU64`
for `rustc_data_structures` and rewrites the spec for `sparc_unknown_linux_gnu` to use
`TargetOptions` and replaces the previously used `-mv8plus` with the more portable
`-mcpu=v9 -m32`.

To make `rustc` build successfully, support for 32-bit SPARC needs to be added to the `object`
crate as well as the `nix` crate which I will be sending out later as well.

r? nagisa

### CODE REVIEW

**Comment by nagisa** [comment]
Seems reasonable enough. Thank you!

r+

File: compiler/rustc_data_structures/src/marker.rs [hunk 1]
```diff
@@ -147,14 +147,14 @@ cfg_match! {
             [crate::owned_slice::OwnedSlice]
         );
 
-        // PowerPC and MIPS platforms with 32-bit pointers do not
+        // MIPS, PowerPC and SPARC platforms with 32-bit pointers do not
         // have AtomicU64 type.
-        #[cfg(not(any(target_arch = "powerpc", target_arch = "mips")))]
+        #[cfg(not(any(target_arch = "powerpc", target_arch = "powerpc", target_arch = "sparc")))]
```
Comment (roadrunnerto100) [inline compiler/rustc_data_structures/src/marker.rs:152]: You seem to have listed powerpc twice

Comment (glaubitz) [reply compiler/rustc_data_structures/src/marker.rs:152]: Oh shoot, you're right. I will send a follow-up PR to fix this. Thanks a lot for catching this!

Comment (glaubitz) [reply compiler/rustc_data_structures/src/marker.rs:152]: Created https://github.com/rust-lang/rust/pull/127075 now. Again, thanks so much for catching this!

**Pull request merged**